File: js-ts-automation/playwrite/Pages/LoginPage.ts
import { Page, expect } from '@playwright/test';
import { URLS } from '../data/testData';

/**
 * LoginPage — encapsulates all interactions with the SauceDemo login screen.
 * Assertions that belong to the page live here so tests remain concise.
 */
export class LoginPage {
  private readonly page: Page;

  // Locators as readonly fields — one place to update if selectors change
  readonly usernameInput = () => this.page.getByTestId('username');
  readonly passwordInput = () => this.page.getByTestId('password');
  readonly loginButton = () => this.page.getByTestId('login-button');
  readonly errorMessage = () => this.page.getByTestId('error');

  constructor(page: Page) {
    this.page = page;
  }

  /** Navigate directly to the login page */
  async goto(): Promise<void> {
    await this.page.goto(URLS.base);
  }

  /** Fill credentials and click Login */
  async login(username: string, password: string): Promise<void> {
    await this.usernameInput().fill(username);
    await this.passwordInput().fill(password);
    await this.loginButton().click();
  }

  /** Attempt login then assert the expected error text is displayed */
  async loginAndExpectError(username: string, password: string, errorText: string): Promise<void> {
    await this.login(username, password);
    await expect(this.errorMessage()).toBeVisible();
    await expect(this.errorMessage()).toContainText(errorText);
  }

  /** Assert the page landed on the inventory after successful login */
  async assertSuccessfulLogin(): Promise<void> {
    await expect(this.page).toHaveURL(URLS.inventory);
  }

  /** Assert the X icon exists on the error banner (visual cue for recruiters) */
  async assertErrorIconVisible(): Promise<void> {
    await expect(this.errorMessage().locator('svg[data-icon="times"]')).toBeVisible();
  }
}

File: reqnroll-automation/playwright/Pages/LoginPage.cs
using Microsoft.Playwright;
using NUnit.Framework;
using ReqnrollPlaywright.Fixtures;

namespace ReqnrollPlaywright.Pages;

/// <summary>
/// LoginPage — encapsulates all interactions with the SauceDemo login screen.
/// Assertions that belong to the page live here so tests remain concise.
/// </summary>
public class LoginPage
{
    private readonly IPage _page;

    public LoginPage(IPage page)
    {
        _page = page;
    }

    public ILocator UsernameInput => _page.Locator("[data-test='username']");
    public ILocator PasswordInput => _page.Locator("[data-test='password']");
    public ILocator LoginButton => _page.Locator("[data-test='login-button']");
    public ILocator ErrorMessage => _page.Locator("[data-test='error']");

    /// <summary>Navigate directly to the login page.</summary>
    public async Task GoToAsync() => await _page.GotoAsync(Urls.Base);

    /// <summary>Fill credentials and click Login.</summary>
    public async Task LoginAsync(string username, string password)
    {
        await UsernameInput.FillAsync(username);
        await PasswordInput.FillAsync(password);
        await LoginButton.ClickAsync();
    }

    /// <summary>Attempt login then assert the expected error text is displayed.</summary>
    public async Task LoginAndExpectErrorAsync(string username, string password, string expectedText)
    {
        await LoginAsync(username, password);
        Assert.That(await ErrorMessage.IsVisibleAsync(), Is.True);
        Assert.That(await ErrorMessage.TextContentAsync(), Does.Contain(expectedText));
    }

    /// <summary>Assert the page landed on the inventory after successful login.</summary>
    public Task AssertLoggedInAsync()
    {
        Assert.That(_page.Url, Is.EqualTo(Urls.Inventory));
        return Task.CompletedTask;
    }

    /// <summary>Assert the X icon exists on the error banner (visual cue for recruiters).</summary>
    public async Task AssertErrorIconVisibleAsync()
    {
        Assert.That(await ErrorMessage.Locator("svg[data-icon='times']").IsVisibleAsync(), Is.True);
    }
}
